File: backend/app/schemas/market_data.py
"""
Pydantic schemas for market data.
"""
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field

from app.models.market_data import AssetClass


# Base schemas
class AssetBase(BaseModel):
    """Base schema for an asset."""
    symbol: str
    name: str
    asset_class: AssetClass
    description: Optional[str] = None


class MarketDataBase(BaseModel):
    """Base schema for market data."""
    timestamp: datetime
    open_price: Optional[float] = None
    high_price: Optional[float] = None
    low_price: Optional[float] = None
    close_price: float
    volume: Optional[float] = None
    source: str


class EconomicIndicatorBase(BaseModel):
    """Base schema for economic indicators."""
    name: str
    value: float
    region: str
    timestamp: datetime
    source: str


class MarketMetricBase(BaseModel):
    """Base schema for market metrics."""
    name: str
    value: float
    timestamp: datetime


# Create schemas
class AssetCreate(AssetBase):
    """Schema for creating an asset."""
    pass


class MarketDataCreate(MarketDataBase):
    """Schema for creating market data."""
    asset_id: int


class EconomicIndicatorCreate(EconomicIndicatorBase):
    """Schema for creating an economic indicator."""
    pass


class MarketMetricCreate(MarketMetricBase):
    """Schema for creating a market metric."""
    asset_id: Optional[int] = None


# Response schemas
class Asset(AssetBase):
    """Schema for asset response."""
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        """Pydantic configuration."""
        from_attributes = True


class MarketData(MarketDataBase):
    """Schema for market data response."""
    id: int
    asset_id: int
    created_at: datetime

    class Config:
        """Pydantic configuration."""
        from_attributes = True


class EconomicIndicator(EconomicIndicatorBase):
    """Schema for economic indicator response."""
    id: int
    created_at: datetime

    class Config:
        """Pydantic configuration."""
        from_attributes = True


class MarketMetric(MarketMetricBase):
    """Schema for market metric response."""
    id: int
    asset_id: Optional[int] = None
    created_at: datetime

    class Config:
        """Pydantic configuration."""
        from_attributes = True


# List response schemas
class AssetList(BaseModel):
    """Schema for list of assets."""
    items: List[Asset]
    total: int


class MarketDataList(BaseModel):
    """Schema for list of market data points."""
    items: List[MarketData]
    total: int


class EconomicIndicatorList(BaseModel):
    """Schema for list of economic indicators."""
    items: List[EconomicIndicator]
    total: int


class MarketMetricList(BaseModel):
    """Schema for list of market metrics."""
    items: List[MarketMetric]
    total: int


# Dashboard specific schemas
class AssetWithLatestData(BaseModel):
    """Schema for asset with latest market data."""
    asset: Asset
    latest_data: Optional[MarketData] = None
    metrics: List[MarketMetric] = []

    class Config:
        """Pydantic configuration."""
        from_attributes = True


class MarketDataTimeSeriesPoint(BaseModel):
    """Schema for a single point in a time series."""
    timestamp: datetime
    value: float


class MarketDataTimeSeries(BaseModel):
    """Schema for market data time series."""
    asset_id: int
    asset_symbol: str
    asset_name: str
    metric: str
    data: List[MarketDataTimeSeriesPoint]


class DashboardSummary(BaseModel):
    """Schema for dashboard summary."""
    assets_by_class: dict
    latest_economic_indicators: List[EconomicIndicator]
    trending_assets: List[AssetWithLatestData]


class MarketDataFilter(BaseModel):
    """Schema for filtering market data."""
    asset_class: Optional[AssetClass] = None
    symbols: Optional[List[str]] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    metrics: Optional[List[str]] = None
    limit: int = 100
    offset: int = 0

File: frontend/src/types/market-data.ts
/**
 * Enum for asset classes
 */
export enum AssetClass {
  EQUITY = "equity",
  FIXED_INCOME = "fixed_income",
  DERIVATIVE = "derivative",
  CURRENCY = "currency",
  COMMODITY = "commodity"
}

/**
 * Asset interface
 */
export interface Asset {
  id: number;
  name: string;
  symbol: string;
  description?: string;
  asset_class: AssetClass;
  region?: string;
  sector?: string;
  industry?: string;
  created_at: string;
  updated_at: string;
}

/**
 * Market data interface
 */
export interface MarketData {
  id: number;
  asset_id: number;
  timestamp: string;
  open_price?: number;
  high_price?: number;
  low_price?: number;
  close_price?: number;
  volume?: number;
  created_at: string;
  updated_at: string;
  source: string;
}

/**
 * Economic indicator interface
 */
export interface EconomicIndicator {
  id: number;
  name: string;
  value: number;
  timestamp: string;
  region: string;
  description?: string;
  created_at: string;
  updated_at: string;
  source: string;
}

/**
 * Market metric interface
 */
export interface MarketMetric {
  id: number;
  asset_id: number;
  name: string;
  value: number;
  timestamp: string;
  created_at: string;
  updated_at: string;
}

/**
 * Asset with latest data and metrics
 */
export interface AssetWithLatestData {
  asset: Asset;
  latest_data?: MarketData;
  metrics: MarketMetric[];
}

/**
 * Dashboard summary interface
 */
export interface DashboardSummary {
  trending_assets: AssetWithLatestData[];
  latest_economic_indicators: EconomicIndicator[];
}

/**
 * Asset list response
 */
export interface AssetList {
  items: Asset[];
  total: number;
}

/**
 * Market data list response
 */
export interface MarketDataList {
  items: MarketData[];
  total: number;
}

/**
 * Economic indicator list response
 */
export interface EconomicIndicatorList {
  items: EconomicIndicator[];
  total: number;
}

/**
 * Market metric list response
 */
export interface MarketMetricList {
  items: MarketMetric[];
  total: number;
}

/**
 * Time series data point
 */
export interface TimeSeriesDataPoint {
  timestamp: string;
  value: number;
}

/**
 * Market data time series response
 */
export interface MarketDataTimeSeries {
  asset_id: number;
  asset_symbol: string;
  metric: string;
  data: TimeSeriesDataPoint[];
}

/**
 * Market data filter options
 */
export interface MarketDataFilter {
  assetClass?: AssetClass;
  region?: string;
  sector?: string;
  startDate?: string;
  endDate?: string;
}
